### ISSUE TITLE
Set created from tuple literal has tuple typing semantics

### ISSUE THREAD

**Issue by layday** [microsoft/pyright#1249]
**Describe the bug**
When creating a set from a tuple literal, the type of the tuple is inherited verbatim by the set, which restricts the types the set will accept to the `Literal` types of the tuple.

**To Reproduce**
Type check the snippet below.

**Expected behavior**
That line 13 would not generate an error:

```
Argument of type "Iterable[str]" cannot be assigned to parameter "s" of type "Iterable[Literal['']]" in function "union"
  TypeVar "_T_co" is covariant
    "str" cannot be assigned to type "Literal['']"
```

**Screenshots or Code**
```python
# pyright: strict

from __future__ import annotations

from collections.abc import Iterable

def foo() -> Iterable[str]:
    ...

set({''}).union(foo())
set(['']).union(foo())
set(('',)).union(foo())

```

**VS Code extension or command-line**
Either, 1.1.93

**Comment by erictraut** [comment]
This behavior is "as designed". What you're seeing here is not specific to a set. The same behavior occurs for lists created from tuples with literals, for example.

The current behavior is the natural result of the tuple type inference rules documented [here](https://github.com/microsoft/pyright/blob/master/docs/type-inference.md#tuple-expressions).

Here's something we could consider. The tuple is special in that it supports variadic type parameters. Yet, its class definition in builtins.pyi indicates that it derives from `Sequence[_T_co]`, and `_T_co` is not variadic. That means the type checker needs to combine the variadic type arguments into a single type argument. It does so by union'ing them together. It would be possible for the type checker to drop the literals when performing this union. The question is whether that's always the desired behavior. What do you think?

**Comment by layday** [comment]
I think that's reasonable and it would behave identically to a set or list literal, as I understanding it, i.e. `reveal_type(set(('',))) == reveal_type({''})` would be true, which is what one might expect.

**Comment by erictraut** [comment]
Alright, I've made the behavior change. This will be in the next release.

**Comment by erictraut** [comment]
This is addressed in Pyright 1.1.94, which I just published. It will also be included in this week's Pylance release.
